<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> Hierarchical Fusion Evolving Spiking Neural Network for Adaptive Learning </title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Analysis of Massive SNP Arrays">
<meta property="og:description" content="Easy-to-use, efficient, flexible and scalable tools for analyzing 
    massive SNP arrays. PrivÃ© et al. (2018) &lt;doi:10.1093/bioinformatics/bty185&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">HeSNN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="articles/exo.html">Exercise</a>
    </li>

  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://www.obadalazoubi.com">
    <span class="fas fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/obada-alzoubi/HF_eSNN">
    <span class="fas fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="news/index.html">
    <span class="fas fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div id="HeSNN" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#bigsnpr" class="anchor"></a>HeSNN</h1></div>
<p>{HeSNN} is matlab toolbox for implementing Evolving Spiking Neural Networks.</p>
<p> </p>A majority of machine learning (ML) approaches functions in offline or batch modes, which limits their application to adaptive environments. 
Thus, developing algorithms that work in adaptive and dynamic environments is the subject of ongoing research. Such algorithms require to learn not only from new samples (online learning), but also from novel and unseen before knowledge. </p> 
<p> Here, the toolbox uses evolving learning (EL) to refer to learning from new knowledge and unseen-before classes without needing to re-train models as in traditional ML methods. </p>
 To achieve the goal of EL, we adopt a biologically-inspired paradigm to build a highly adaptive supervised learning algorithm based on two brain-like information processing: divide-and-conquer and hierarchical abstraction. </p>
<p> Furthermore, our proposed algorithm, which we named it as Hierarchical Fusion Evolving Spiking Neural Network (HFSNN), uses a dynamical and biologically inspired spiking neural network (SNN) with the optimized neural model. </p>
<p> HFSNN does not impose any limitation on the data regarding the number of classes or the way of feeding the data to the model. The toolbox implements several features including learning in offline, online and evolving learning mods and establish for future applications for EL.

<p><a href="https://privefl.github.io/bigsnpr/articles/demo.html">Quick demo</a></p>
<p><strong>Note the toolbox is highly customizable and targets development and building eSNN</strong> </p>
<p><strong>Currently, HeSNN supports classification only. </strong> </p>

<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>In Matlab, run</p>


<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="op">
% clc;clear 
% Add path for some needed codes 
addpath('matlab') 
ddpath('data')
addpath('coreCode')
</span>
</pre></div>

<div id="input-formats" class="section level2">
<h2 class="hasAnchor">
<a href="#input-formats" class="anchor"></a> Feature Encoding </h2>
<p>The rank coding procedure is the following: first, it starts by converting the input values into a sequence of spikes using the Gaussian receptive fields. The Gaussian receptive fields consist of m receptive fields that are used to represent the input value n into spikes. Assume that n takes values from the range [I_min^n  ,I_max^n], then the Gaussian receptive field of neuron I is given by its center u_i:  
HeSNN uses population encoding to convert features into set of spikes. The goal is to convert each feature to <code>nbfields</code>
neurons. 
</p>
<p><img src="reference/eq1.png" width="200" style="display: block; margin: auto;"></p>
<p>with </p>
<p><img src="reference/eq2.png" width="120" style="display: block; margin: auto;"></p>

<p><img src="reference/grf_population_encoding.png" width="700" style="display: block; margin: auto;"></p>

<a href="#input-formats" class="anchor"></a>Input formats</h2>

<p> HeSNN uses matrix format to represent the data with last column is the class label.
The features should be scaled between -1 and +1 in order to have correct encoding for the features. </p>
<ul>
<li>
<code>Data</code>: A matrix <code>(nxm)</code>, with m column being corresponding labels. </li>
<li>
<code>Param</code>: A structure with several fields to control the training process. 
<ul>
<li> <code> Beta:</code> the width of the  </li>
<li> <code> I_min:</code> the lower level of receptive filed. </li>
<li> <code> I_max:</code> the upper level of receptive filed. </li>
<li> <code> nbfields:</code> the number of receptive fields per feature.  </li>
<li> <code> max_response_threshold:</code> firing threshold for neurons. </li>
<li> <code> s:</code> merging threshold for neurons.</li>
<li> <code> m:</code> exponential base decaying factors for neurons. </li>
<li> <code> c:</code> discounting threshold for neurons. </li>
<li> <code> distance:</code> the distance measure to merge neurons. </li>
<li> <code> distribution:</code> the distribution for the receptive filed  </li>

<li> <code> Beta:</code> the width of the  </li>

</ul>

</li>
<li>



<div id="polygenic-scores" class="section level2">
<h2 class="hasAnchor">
<a href="#polygenic-scores" class="anchor"></a>Demo Example</h2>
<p>In this demo example, we will use fisheriris dataset as an exmaple:</p>
<ul>
<li><p>Penalized regressions with individual-level data (see <a href="https://doi.org/10.1534/genetics.119.302019">paper</a> and <a href="https://privefl.github.io/bigstatsr/articles/penalized-regressions.html">tutorial</a>)</p></li>
<li><p>Clumping and Thresholding (C+T) and Stacked C+T (SCT) with summary statistics and individual level data (see <a href="https://doi.org/10.1016/j.ajhg.2019.11.001">paper</a> and <a href="https://privefl.github.io/bigsnpr/articles/SCT.html">tutorial</a>).</p></li>
<li><p>LDpred2 with summary statistics (see <a href="https://doi.org/10.1101/2020.04.28.066720">preprint</a> and <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">tutorial</a>)</p></li>
</ul>
</div>
<div id="possible-upcoming-features" class="section level2">
<h2 class="hasAnchor">
<a href="#possible-upcoming-features" class="anchor"></a>Possible upcoming features</h2>
<ul>
<li><p>Multiple imputation for GWAS (https://doi.org/10.1371/journal.pgen.1006091).</p></li>
<li><p>More interactive (visual) QC.</p></li>
</ul>
<p>You can request some feature by opening an issue.</p>
</div>
<div id="bug-report" class="section level2">
<h2 class="hasAnchor">
<a href="#bug-report" class="anchor"></a>Bug report</h2>
<p><a href="https://stackoverflow.com/q/5963269/6103040">How to make a great R reproducible example?</a></p>
<p>Please open an issue if you find a bug.</p>
<p>I will always redirect you to GitHub issues if you email me, so that others can benefit from our discussion.</p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<ul>
<li><p>PrivÃ©, Florian, et al.Â <a href="https://doi.org/10.1101/2020.04.28.066720">âLDpred2: better, faster, stronger.â</a> BioRxiv (2020).</p></li>
</ul>
<p><br></p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=bigsnpr">https://âcloud.r-project.org/âpackage=bigsnpr</a>
</li>
<li>Browse source code at <br><a href="https://github.com/privefl/bigsnpr/">https://âgithub.com/âprivefl/âbigsnpr/â</a>
</li>
<li>Report a bug at <br><a href="https://github.com/privefl/bigsnpr/issues">https://âgithub.com/âprivefl/âbigsnpr/âissues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing bigsnpr</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Florian PrivÃ© <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/privefl/bigsnpr/actions"><img src="https://github.com/privefl/bigsnpr/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://ci.appveyor.com/project/privefl/bigsnpr"><img src="https://ci.appveyor.com/api/projects/status/github/privefl/bigsnpr?branch=master&amp;svg=true" alt="AppVeyor Build Status"></a></li>
<li><a href="https://codecov.io/github/privefl/bigsnpr"><img src="https://img.shields.io/codecov/c/github/privefl/bigsnpr/master.svg" alt="Coverage Status"></a></li>
<li><a href="https://CRAN.R-project.org/package=bigsnpr"><img src="http://www.r-pkg.org/badges/version/bigsnpr" alt="CRAN_Status_Badge"></a></li>
<li><a href="http://dx.doi.org/10.1093/bioinformatics/bty185"><img src="https://zenodo.org/badge/doi/10.1093/bioinformatics/bty185.svg" alt="DOI"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Obada Al Zoubi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
